### PR TITLE
drivers: flash: stm32l4x: Support SOCs with < 1MB

### PR DESCRIPTION
**Pull request by gon1332**
Up to now, the only SOC that was supported in the STM32L496 series was
STM32L496XG with flash of 1MB. With the recent support of STM32L496XE,
this implementation is not correct. This patch adds support for the
other SOCs that come with flashes of 512KB and 256KB.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

### CODE REVIEW

**Comment by gon1332** [comment]
I compared my changes against the L476xx, L496xx, L432xx, L433xx, L452xx, L475xx, L471xx and L4R5xx reference manuals.

The changes seem OK with them. Could anyone else also check it?

**Comment by gon1332** [comment]
Erase works after increasing the max wait time, as described in https://github.com/zephyrproject-rtos/zephyr/issues/16934.

**Review by erwango — APPROVED** [review]

File: drivers/flash/flash_stm32l4x.c [hunk 1]
```diff
@@ -114,8 +115,21 @@ static int erase_page(struct device *dev, unsigned int page)
...
 	struct stm32l4x_flash *regs = FLASH_STM32_REGS(dev);
 	u32_t tmp;
+	u16_t pages_per_bank;
 	int rc;
 
+#if DT_FLASH_SIZE == 1024
+	/* The SOC has 1MB of FLASH with DUALBANKing by design */
+	pages_per_bank = 256U;
+#elif DT_FLASH_SIZE < 1024
+#ifndef FLASH_OPTR_DUALBANK
```
Comment (galak) [inline drivers/flash/flash_stm32l4x.c]: Can you make this:

```
#ifdef FLASH_OPTR_DUALBANK
...
#else 
...
#endif 
```

**Review by galak — CHANGES_REQUESTED** [review]
Minor style change, make it #ifdef FLASH_OPTR_DUALBANK instead of #ifndef

File: drivers/flash/flash_stm32l4x.c (hunk 1 above)
Comment (gon1332) [reply drivers/flash/flash_stm32l4x.c]: Done

**Review by galak — APPROVED** [review]

**Pull request merged**